cryptography/src/darwinMain/kotlin/com/wire/kalium/cryptography/utils/CryptoUtils.kt

```diff
@@ -18,81 +18,162 @@
 
 package com.wire.kalium.cryptography.utils
 
-import com.wire.kalium.cryptography.kaliumLogger
-import io.ktor.util.encodeBase64
-import kotlinx.cinterop.allocArrayOf
+import com.wire.kalium.cryptography.exceptions.CryptographyException
+import kotlinx.cinterop.ULongVar
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.alloc
 import kotlinx.cinterop.memScoped
-import kotlinx.cinterop.refTo
-import okio.HashingSink
+import kotlinx.cinterop.ptr
+import kotlinx.cinterop.usePinned
+import kotlinx.cinterop.value
+import okio.Buffer
 import okio.Sink
 import okio.Source
-import okio.blackholeSink
 import okio.buffer
 import okio.use
-import platform.CoreCrypto.CC_MD5
-import platform.CoreCrypto.CC_MD5_DIGEST_LENGTH
-import platform.Foundation.NSData
-import platform.Foundation.base64Encoding
-import platform.Foundation.create
-
-actual fun calcMd5(bytes: ByteArray): String {
-    val digestData = UByteArray(CC_MD5_DIGEST_LENGTH)
-    val data = toData(bytes)
-    CC_MD5(data.bytes, data.length.toUInt(), digestData.refTo(0))
-
-    return toData(digestData.asByteArray()).base64Encoding()
+import platform.CoreCrypto.CCCrypt
+import platform.CoreCrypto.kCCAlgorithmAES
+import platform.CoreCrypto.kCCBlockSizeAES128
+import platform.CoreCrypto.kCCDecrypt
+import platform.CoreCrypto.kCCEncrypt
+import platform.CoreCrypto.kCCKeySizeAES256
+import platform.CoreCrypto.kCCOptionPKCS7Padding
+import platform.CoreCrypto.kCCSuccess
+import platform.Security.SecRandomCopyBytes
+import platform.Security.errSecSuccess
+import platform.Security.kSecRandomDefault
+
+actual fun encryptDataWithAES256(data: PlainData, key: AES256Key): EncryptedData {
+    val outputBuffer = Buffer()
+    val inputBuffer = Buffer()
+
+    return outputBuffer.use { output ->
+        inputBuffer.use { input ->
+            input.write(data.data)
+            encryptFileWithAES256(input, key, output)
+        }
+        EncryptedData(output.readByteArray())
+    }
 }
 
-actual fun calcSHA256(bytes: ByteArray): ByteArray {
-    TODO("Not yet implemented")
+actual fun decryptDataWithAES256(data: EncryptedData, secretKey: AES256Key): PlainData {
+    val outputBuffer = Buffer()
+    val inputBuffer = Buffer()
+
+    return outputBuffer.use { output ->
+        inputBuffer.use { input ->
+            input.write(data.data)
+            decryptFileWithAES256(input, output, secretKey)
+        }
+        PlainData(output.readByteArray())
+    }
 }
 
-@Suppress("TooGenericExceptionCaught")
-actual fun calcFileMd5(dataSource: Source): String? =
+actual fun encryptFileWithAES256(assetDataSource: Source, key: AES256Key, outputSink: Sink): Long {
     try {
-        dataSource.buffer().use { source ->
-            HashingSink.md5(blackholeSink()).use { sink ->
-                source.readAll(sink)
-                sink.hash.toByteArray().encodeBase64()
+        val iv = generateRandomData(kCCBlockSizeAES128.toInt())
+        val plainData = assetDataSource.buffer().readByteArray()
```

Comment:
Did you try this on big asset files?

Reply:
No I didn't try it on big files but I did add TODO for using streaming or blocked based API so that we don't have to read the whole file at once.